src/lightning/pytorch/loops/prediction_loop.py

```diff
@@ -201,41 +205,57 @@ def teardown(self) -> None:
             self._data_fetcher.teardown()
             self._data_fetcher = None
 
-    def _predict_step(self, batch: Any, batch_idx: int, dataloader_idx: int) -> None:
+    def _predict_step(
+        self, batch: Any, batch_idx: int, dataloader_idx: int, dataloader_iter: Optional[Iterator]
+    ) -> None:
         """Runs the actual predict step together with all the necessary bookkeeping and the hooks tied to it.
 
         Args:
             batch: the current batch to run the prediction on
-            batch_idx: the index of the current batch
-            dataloader_idx: the index of the dataloader producing the current batch
+            batch_idx: The index of the current batch.
+            dataloader_idx: the index of the dataloader producing the current batch.
+            dataloader_iter: The iterator if using this step flavor.
 
         """
         trainer = self.trainer
-        batch = trainer.precision_plugin.convert_input(batch)
-        batch = trainer.lightning_module._on_before_batch_transfer(batch, dataloader_idx=dataloader_idx)
-        batch = call._call_strategy_hook(trainer, "batch_to_device", batch, dataloader_idx=dataloader_idx)
+        assert (data_fetcher := self._data_fetcher) is not None
+
+        if not (using_dataloader_iter := isinstance(data_fetcher, _DataLoaderIterDataFetcher)):
+            batch = trainer.precision_plugin.convert_input(batch)
+            batch = trainer.lightning_module._on_before_batch_transfer(batch, dataloader_idx=dataloader_idx)
+            batch = call._call_strategy_hook(trainer, "batch_to_device", batch, dataloader_idx=dataloader_idx)
 
         self.batch_progress.increment_ready()
 
-        any_on_epoch = self._store_data_for_prediction_writer(batch_idx, dataloader_idx)
+        if not using_dataloader_iter:
+            any_on_epoch = self._store_data_for_prediction_writer(batch_idx, dataloader_idx)
 
-        step_kwargs = self._build_kwargs(batch, batch_idx, dataloader_idx if self.num_dataloaders > 1 else None)
+        # the `_step` methods don't take a batch_idx when `dataloader_iter` is used, but all other hooks still do,
+        # so we need different kwargs
+        hook_kwargs = self._build_kwargs(batch, batch_idx, dataloader_idx if self.num_dataloaders > 1 else None)
+        step_kwargs = hook_kwargs if not using_dataloader_iter else {"any": dataloader_iter}
```

Comment:
Below we are converting the kwargs to args, so maybe simplify this to just an args tuple from the start?